### PR TITLE
add size parameter for `vec_box`  lint

### LINKED ISSUE #3547: Make VEC_BOX lint configurable over the size of T
In #3545 the `VEC_BOX` lint was implemented. This lints on usage of `Vec<Box<T: Sized>>` . The `Box` around `T` could make sense though, if `T` is a very large type. (See also https://github.com/rust-lang/rust-clippy/issues/3530#issuecomment-446058452)

A good enhancement for this lint would be, to make it configurable over the size of `T`, with some meaningful default value (TBD).

### PR DESCRIPTION
**Pull request by Areredify**
changelog: add size threshold for the `vec_box` lint, currently 4096 bytes (one page) (subject to change). Closes #3547.

diff is a little bit confusing due to some refactoring (moving free functions to lint struct functions), relevant portion is [this](https://github.com/rust-lang/rust-clippy/compare/master...Areredify:vec_box_threshold?expand=1#diff-1096120ca9143af89dcc9175ea92b54aR294-R298). In hindsight should've been different commits, but oh well.

### CODE REVIEW

File: tests/ui-toml/toml_unknown_key/conf_unknown_key.stderr [hunk 1]
```diff
@@ -1,4 +1,4 @@
-error: error reading Clippy's configuration file `$DIR/clippy.toml`: unknown field `foobar`, expected one of `blacklisted-names`, `cognitive-complexity-threshold`, `cyclomatic-complexity-threshold`, `doc-valid-idents`, `too-many-arguments-threshold`, `type-complexity-threshold`, `single-char-binding-names-threshold`, `too-large-for-stack`, `enum-variant-name-threshold`, `enum-variant-size-threshold`, `verbose-bit-mask-threshold`, `literal-representation-threshold`, `trivial-copy-size-limit`, `too-many-lines-threshold`, `array-size-threshold`, `third-party` at line 5 column 1
+error: error reading Clippy's configuration file `$DIR/clippy.toml`: unknown field `foobar`, expected one of `blacklisted-names`, `cognitive-complexity-threshold`, `cyclomatic-complexity-threshold`, `doc-valid-idents`, `too-many-arguments-threshold`, `type-complexity-threshold`, `single-char-binding-names-threshold`, `too-large-for-stack`, `enum-variant-name-threshold`, `enum-variant-size-threshold`, `verbose-bit-mask-threshold`, `literal-representation-threshold`, `trivial-copy-size-limit`, `too-many-lines-threshold`, `array-size-threshold`, `vec-box-size-threshold`, `third-party` at line 5 column 1
```
Comment (flip1995) [inline tests/ui-toml/toml_unknown_key/conf_unknown_key.stderr]: Please also add a ui_toml test, where you set the size config.

Comment (Areredify) [reply tests/ui-toml/toml_unknown_key/conf_unknown_key.stderr]: I don't understand which test do you mean

Comment (flip1995) [reply tests/ui-toml/toml_unknown_key/conf_unknown_key.stderr]: A test like this: https://github.com/rust-lang/rust-clippy/tree/9f858e1acbf39c4a93152ea0f4b3652d36fe90a6/tests/ui-toml/functions_maxlines

Comment (Areredify) [reply tests/ui-toml/toml_unknown_key/conf_unknown_key.stderr]: all good now (I hope)

File: clippy_lints/src/utils/conf.rs [hunk 2]
```diff
@@ -152,6 +152,8 @@ define_Conf! {
     (too_many_lines_threshold, "too_many_lines_threshold", 100 => u64),
     /// Lint: LARGE_STACK_ARRAYS. The maximum allowed size for arrays on the stack
     (array_size_threshold, "array_size_threshold", 512_000 => u64),
+    /// Lint: VEC_BOX. The minimum required size for boxed type in Vec
```
Comment (flip1995) [inline clippy_lints/src/utils/conf.rs]:
```suggestion
    /// Lint: VEC_BOX. The size of the boxed type in bytes, where boxing in a `Vec` is allowed
```

Is 4kB based on something?

Comment (Areredify) [reply clippy_lints/src/utils/conf.rs]: Page size; but as I said in the pr, you are welcome to put any number you like

Comment (flip1995) [reply clippy_lints/src/utils/conf.rs]: Page size sounds good.

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Pull request merged**